File: internal/protocol/protocol_test.go
package protocol

import "testing"

func TestKeysAndSequences(t *testing.T) {
	if got := OpenKey("/demo/", "abc"); got != "demo/v1/open/abc.json" {
		t.Fatalf("OpenKey = %q", got)
	}
	if got := DataKey("demo", "c2s", "abc", 42); got != "demo/v1/data/c2s/abc/00000000000000000042.bin" {
		t.Fatalf("DataKey = %q", got)
	}
	if got := AckKey("demo", "s2c", "abc"); got != "demo/v1/ack/s2c/abc.json" {
		t.Fatalf("AckKey = %q", got)
	}
	seq, err := ParseSeq("demo/v1/data/c2s/abc/00000000000000000042.bin")
	if err != nil || seq != 42 {
		t.Fatalf("ParseSeq = %d, %v", seq, err)
	}
}

func TestChunkSizeValidationAndEffectiveSendLimit(t *testing.T) {
	for _, bad := range []int{0, -1, MinChunkSize - 1, MaxChunkSize + 1} {
		if err := ValidateChunkSize(bad); err == nil {
			t.Fatalf("ValidateChunkSize(%d) succeeded", bad)
		}
	}
	for _, good := range []int{MinChunkSize, 64 * 1024, MaxChunkSize} {
		if err := ValidateChunkSize(good); err != nil {
			t.Fatalf("ValidateChunkSize(%d): %v", good, err)
		}
	}
	got, err := EffectiveSendChunkSize(64*1024, 32*1024)
	if err != nil {
		t.Fatal(err)
	}
	if got != 32*1024 {
		t.Fatalf("effective = %d", got)
	}
	got, err = EffectiveSendChunkSize(32*1024, 64*1024)
	if err != nil {
		t.Fatal(err)
	}
	if got != 32*1024 {
		t.Fatalf("effective = %d", got)
	}
}

func TestTargetIPv6JSON(t *testing.T) {
	target := Target{Type: AddressIPv6, Host: "2001:db8::1", Port: 443}
	data, err := Marshal(target)
	if err != nil {
		t.Fatal(err)
	}
	var got Target
	if err := Unmarshal(data, &got); err != nil {
		t.Fatal(err)
	}
	if got != target {
		t.Fatalf("target mismatch: %#v", got)
	}
	if got.Address() != "[2001:db8::1]:443" {
		t.Fatalf("IPv6 address formatting = %q", got.Address())
	}
}

File: android-client/app/src/test/java/io/s3s5/android/protocol/ProtocolTest.kt
package io.s3s5.android.protocol

import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test

class ProtocolTest {
    @Test
    fun keysMatchGoLayout() {
        val session = "0123456789abcdef0123456789abcdef"
        assertEquals("s3s5/v1/open/$session.json", Protocol.openKey("", session))
        assertEquals("p/v1/open-result/$session.json", Protocol.openResultKey("/p/", session))
        assertEquals("p/v1/data/c2s/$session/00000000000000000042.bin", Protocol.dataKey("p", "c2s", session, 42))
        assertEquals("p/v1/ack/s2c/$session.json", Protocol.ackKey("p", "s2c", session))
        assertEquals("p/v1/close/server/$session.json", Protocol.closeKey("p", "server", session))
        assertEquals("p/v1/open/", Protocol.openPrefix("p"))
    }

    @Test
    fun sequenceFormattingMatchesGo() {
        assertEquals("00000000000000000000", Protocol.formatSeq(0))
        assertEquals("00000000000000000007", Protocol.formatSeq(7))
        assertEquals(7, Protocol.parseSeq("s3s5/v1/data/c2s/id/00000000000000000007.bin"))
    }

    @Test
    fun sessionIdsAreThirtyTwoHexChars() {
        val id = Protocol.newSessionId()
        assertTrue(id.matches(Regex("[0-9a-f]{32}")))
    }

    @Test
    fun chunkLimitsAreValidated() {
        Protocol.validateChunkSize(MIN_CHUNK_SIZE)
        Protocol.validateChunkSize(MAX_CHUNK_SIZE)
        assertEquals(4096, Protocol.effectiveSendChunkSize(8192, 4096))
        assertEquals(4096, Protocol.effectiveSendChunkSize(4096, 8192))
    }
}
